File: Python3/801 - 1000/870 Advantage Shuffle.py
# 580ms, 12.88%; 15.9MB, 83.33%
class Solution:
    def advantageCount(self, A: list, B: list) -> list:
        res, tmp = [], sorted(A)
        for b in B:
            idx = self.upper_bound(tmp, b)
            res.append(tmp.pop(idx))
        return res

    def upper_bound(self, arr: list, target: int) -> int:
        if arr[-1] <= target: return 0
        beg, end = 0, len(arr) - 1
        while beg < end:
            mid = (beg + end) // 2
            if arr[mid] > target:
                end = mid
            else:
                beg = mid + 1
        return beg

File: 801 - 1000/870. Advantage Shuffle.cc
// 172ms, 73.39%; 21.5MB, 50.00%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> advantageCount(const vector<int> &A, const vector<int> &B) {
        vector<int> res(A.size());
        multiset<int> tmp(A.begin(), A.end());
        for (int i = 0, size = B.size(); i < size; ++i) {
            auto iter = tmp.upper_bound(B[i]);
            if (iter == tmp.end()) iter = tmp.begin();
            res[i] = *iter;
            tmp.erase(iter);
        }
        return res;
    }
};
